types/wlr_surface.c

```diff
@@ -91,11 +91,33 @@ static void surface_set_input_region(struct wl_client *client,
 	}
 }
 
+static void wlr_surface_get_effective_size(struct wlr_surface *surface,
+		int swidth, int sheight, int *width, int *height) {
+	int scale = surface->current.scale;
+	enum wl_output_transform transform = surface->current.transform;
+	int _width = swidth / scale;
+	int _height = sheight / scale;
```

Comment:
Yeah you need this division here for it to work correctly.

We could get this a bit nicer with a little refactoring of the surface implementation and texture interface and I think that's worth doing eventually. Maybe I'll make an issue for it.

If it were up to me, I would cache this on surface.attach. That's what the letter of the spec says to do.